website/src/routes/api/(async)/optionalAsync/index.mdx

```diff
@@ -1,10 +1,235 @@
 ---
 title: optionalAsync
+description: Creates an optional schema.
 source: /schemas/optional/optionalAsync.ts
 contributors:
   - fabian-hiller
+  - santoshyadavdev
+  - EltonLobo07
 ---
 
+import { Link } from '@builder.io/qwik-city';
+import { ApiList, Property } from '~/components';
+import { properties } from './properties';
+
 # optionalAsync
 
-> The content of this page is not yet ready. Until then, please use the [source code](https://github.com/fabian-hiller/valibot/blob/main/library/src/schemas/optional/optionalAsync.ts) or take a look at [issue #287](https://github.com/fabian-hiller/valibot/issues/287) to help us extend the API reference.
+Creates an optional schema.
+
+```ts
+const Schema = v.optionalAsync<TWrapped, TDefault>(wrapped, default_);
+```
+
+## Generics
+
+- `TWrapped` <Property {...properties.TWrapped} />
+- `TDefault` <Property {...properties.TDefault} />
+
+## Parameters
+
+- `wrapped` <Property {...properties.wrapped} />
+- `default_` {/* prettier-ignore */}<Property {...properties.default_} />
+
+### Explanation
+
+With `optionalAsync` the validation of your schema will pass `undefined` inputs, and if you specify a `default_` input value, the schema will use it if the input is `undefined`. For this reason, the output type may differ from the input type of the schema.
+
+> Note that `optionalAsync` does not accept `null` as an input. If you want to accept `null` inputs, use <Link href="../nullableAsync/">`nullableAsync`</Link>, and if you want to accept `null` and `undefined` inputs, use <Link href="../nullishAsync/">`nullishAsync`</Link> instead. Also, if you want to set a default output value for any invalid input, you should use <Link href="../fallbackAsync/">`fallbackAsync`</Link> instead.
+
+## Returns
+
+- `Schema` <Property {...properties.Schema} />
+
+## Examples
+
+The following examples show how `optionalAsync` can be used.
+
+### Optional username schema
+
+Schema that accepts a unique username or `undefined`.
+
+> By using a function as the `default_` parameter, the schema will return a unique username from the function call each time the input is `undefined`.
+
+```ts
+import { getUniqueUsername, isUsernameUnique } from '~/api';
+
+const OptionalUsernameSchema = v.optionalAsync(
+  v.pipeAsync(
+    v.string(),
+    v.nonEmpty(),
+    v.checkAsync(isUsernameUnique, 'The username is not unique.')
+  ),
+  getUniqueUsername
+);
+```
+
+### New user schema
+
+Schema to validate new user details.
+
+```ts
+const NewUserSchema = v.objectAsync({
+  // Assume the email is checked
```

Comment:
What do you mean by that comment? It is not entirely clear to me.

Reply:
I meant the uniqueness of the email is already checked. This assumption helps to completely skip email async validation like in the example shown below.
```ts
const NewUserSchema = v.objectAsync({
  email: v.pipeAsync(
    v.string(),
    v.email(),
    v.checkAsync(isEmailUnique, 'The email is not unique')
  ),
  username: v.optionalAsync(
    v.pipeAsync(
      v.string(),
      v.nonEmpty(),
      v.checkAsync(isUsernameUnique, 'The username is not unique.')
    )
  ),
  password: v.pipe(v.string(), v.minLength(8)),
});
```
The problem with the above example is that it is sending a wrong message: 'perform backend validation for every field in the object'. I would have restructured my example to this:
```ts
const NewUserSchema = v.pipeAsync(
  v.objectAsync({
    email: v.pipe(v.string(), v.email()),
    username: v.optional(v.pipe(v.string(), v.nonEmpty())),
    password: v.pipe(v.string(), v.minLength(8)),
  }),
  v.partialCheckAsync(
    [['email'], ['username']],
    areUserDetailsUnique,
    'The user details are not unique.'
  )
);
```
But the above example won't demonstrate how to use `optionalAsync`

Do you think I should change the example? If not, what would be your code comment to make the intention clear?

Reply:
I would use this code example:

```ts
const NewUserSchema = v.objectAsync({
  email: v.pipeAsync(
    v.string(),
    v.email(),
    v.checkAsync(isEmailUnique, 'The email is not unique')
  ),
  username: v.optionalAsync(
    v.pipeAsync(
      v.string(),
      v.nonEmpty(),
      v.checkAsync(isUsernameUnique, 'The username is not unique.')
    )
  ),
  password: v.pipe(v.string(), v.minLength(8)),
});
```